lua_cjson.c

```diff
@@ -713,22 +727,39 @@ static void json_append_data(lua_State *l, json_config_t *cfg,
     case LUA_TTABLE:
         current_depth++;
         json_check_encode_depth(l, cfg, current_depth, json);
-        len = lua_array_length(l, cfg, json);
-        if (len > 0 || (len == 0 && !cfg->encode_empty_table_as_object))
+
+        int as_array = 0;
+        int has_metatable = lua_getmetatable(l, -1);
+
+        if (has_metatable) {
+            lua_pushlightuserdata(l, &json_array);
+            lua_rawget(l, LUA_REGISTRYINDEX);
+            as_array = lua_rawequal(l, -1, -2);
+            lua_pop(l, 2);
+        }
+
+        if (as_array) {
+            len = lua_objlen(l, -1);
             json_append_array(l, cfg, current_depth, json, len);
-        else {
-            int as_array = 0;
-            if (lua_getmetatable(l, -1)) {
-                lua_pushlightuserdata(l, &json_empty_array);
-                lua_rawget(l, LUA_REGISTRYINDEX);
-                as_array = lua_rawequal(l, -1, -2);
-                lua_pop(l, 2);
-            }
+        } else {
+            len = lua_array_length(l, cfg, json);
 
-            if (as_array) {
-                json_append_array(l, cfg, current_depth, json, 0);
+            if (len > 0 || (len == 0 && !cfg->encode_empty_table_as_object)) {
+                json_append_array(l, cfg, current_depth, json, len);
             } else {
-                json_append_object(l, cfg, current_depth, json);
+                if (has_metatable) {
+                    lua_getmetatable(l, -1);
+                    lua_pushlightuserdata(l, &json_empty_array);
```

Comment:
Are you sure it's "json_empty_array" instead of "json_array"?

Reply:
Yes, probably. This code was written on top of `master`, and later rebased on top of #26, which ontroduces `json_array`.

Reply:
Had another look at this - this code is correct, it is to preserve backwards compatibility with `cjson.empty_array_mt`.